Load/unload native plugins on a per image basis

AndreasReich opened the pull request:
Today, mono unloads native plugins only on shutdown: Every pinvoke will resolve the necessary dll/so exactly once and if not there put it into a global list which is only cleaned up when we do a complete mono shutdown.
This is problematic for Unity since it bars us from ever reloading plugins. With this PR, all dynamically loaded native plugins are kept in a per assembly-image list. This means of course also, that a native plugin may be "loaded" more than once by different assemblies, thus relying on the OS reference counting to not actually load it more than once.
Due to preloading of embedded libraries in mini/main.c, we keep the global list still around and query that one before looking at the per-image list (this part is new in comparison to the otherwise identical change in the hackweek project last week).
Note that in unity we do a independent dlopen on every native plugin and also never unload it again since this code path is commented out right now. So landing this change alone will have no direct effect on Unity. (which allows us to treat this change completely isolated)


Why do we want this:
* be able to reload native plugins at runtime soon
* properly cleanup things when we expect them to be cleaned up

Why might we not want to do this after all:
* substantial change in mono runtime may be hard to maintain or get upstream
* hard to test on pure mono side? In fact no tests added (all old ones passed though on my mac)
* potential repercussions when going through with the feature on the Unity side
(todo there: uncomment unloading mechanism, look into plugin lifecycle callbacks (broken?!), do copy before load on windows to prevent file locking)

I'm obviously very much in favor of getting this in, but we need to discuss if this is too dangerous/too hard to maintain

Tak committed:
Allow namespace-agnostic class lookup by passing NULL namespace.

vargaz committed:
[jit] Avoid extending the size of Vector2/Vector3 to 16 bytes. Disable SIMD for these types for now, since the SIMD code can't handle types (#6746)

which are smaller than a SIMD reg.

Fixes https://github.com/mono/mono/issues/6411.

richard-fine committed:
Extend mono_profile with more functions to work with coverage data

richard-fine committed:
Fix compile

richard-fine committed:
Change struct definition to a typedef, because this is C

richard-fine committed:
Fix prototype to be compliant C

richard-fine committed:
Free coverage data when unloading a method

richard-fine committed:
Report column information for sequence points when using portable PDBs

richard-fine committed:
Revert free of coverage data, it's in the wrong place

richard-fine committed:
Store a separate coverage hashtable per MonoDomain and init/cleanup when initing/freeing the domain

richard-fine committed:
Expose mono_debug_lookup_source_location_by_il

AndreasReich committed:
UnityTls server no longer throws an exception if there is no client cert during client authentication

AndreasReich committed:
track (most) native modules per image and clean up when image is unloaded

one dlopen per module and image
trust OS to do ref counting.

Tak commented:
Yeah, I think that the temp-copy workaround makes sense

joncham commented on mono/metadata/loader.c:
```diff
@@ -1121,46 +1124,63 @@ cached_module_load (const char *name, int flags, char **err)
-remove_cached_module(gpointer key, gpointer value, gpointer user_data)
+remove_embedded_module (gpointer key, gpointer value, gpointer user_data)
 {
 	mono_dl_close((MonoDl*)value);
```
Since you call `g_strdup` on key (name) above should you free it here?

joncham commented on mono/metadata/image.c:
```diff
@@ -1932,6 +1932,12 @@ mono_image_close_except_pools_all (MonoImage**images, int image_count)
+static void
+remove_cached_module(gpointer key, gpointer value, gpointer user_data)
+{
+	mono_dl_close((MonoDl*)value);
```
Same here are below. Should you `g_free` the key/name?

AndreasReich replied:
I'd say most definitely! Interesting that mono didn't do that anyway. I'll fix it and retest the code.
Do you think running the full mono test suite should be enough to verify such a change? (Alternative would mean that I bring back the hackweek unity reloading hacks (nothing trunkworthy there yet) and manually check against that...)

AndreasReich committed:
Fixed missing g_free on hash table keys

mderoy committed:
Merge pull request #1006 from Unity-Technologies/unity-master-fix-unstable-android

Fix double gchandle free in thread detach logic (1062208)

Commit:
Add option for strict write barriers

Commit:
Add several write barriers needed to comply with strict wbarrier mode

joncham committed:
Apply debugger changes from il2cpp 0c52b0b5b2babc2cbec886b8ef8015f68249026e

mderoy committed:
cleanup the threadinfo before shutdown to fix crash in pthread_key_clean_all

Commit:
Merge pull request #1008 from Unity-Technologies/incremental-boehm-strict-mode

Add option for strict write barriers

joncham committed:
Merge pull request #1009 from Unity-Technologies/unity-master-il2cpp-debugger-sync

Apply debugger changes from il2cpp 0c52b0b5b2babc2cbec886b8ef8015f682…

mderoy committed:
Merge pull request #999 from Unity-Technologies/unity-master-fix-get-offset-initialization

Fix issue where get offset could return 0 when the type is uninited

Commit:
Add another write barriers needed to comply with strict wbarrier mode

Commit:
Fix crash when gc is disabled

Commit:
Merge pull request #1012 from Unity-Technologies/incremental-boehm-strict-mode

Incremental boehm strict mode

Commit:
Make gc params work for Boehm

joncham committed:
Properly inflate pointer types in inflate_generic_type. Generic pointer types are encountered with 'unmanaged' constraint.

Commit:
Merge pull request #1013 from Unity-Technologies/boehm-gc-params

Make gc params work for Boehm

joshpeterson committed:
Build Mono for Android with NDK r16b

Change the compiler used to build Mono for Android from gcc to clang,
and update the NDK to version r16b.

Also, remove some unused ARM architectures.

joshpeterson committed:
Use LF line endings on all perl build scripts

joncham committed:
Don't try to access metadata for dynamic method added to non-dynamic image. Fixes issue #10201.

joshpeterson committed:
Merge pull request #1015 from Unity-Technologies/buildscripts-lf

Use LF line endings on all perl build scripts

joncham committed:
Merge pull request #1016 from Unity-Technologies/unity-master-bug-1069236

Properly inflate pointer types in inflate_generic_type (case 1069236)

joncham committed:
Merge pull request #1017 from Unity-Technologies/unity-master-bug-1065895

Don't try to access metadata for dynamic method (case 1065895)

AndreasReich committed:
[tls] Fixed ignoring error code when numBytesRead/Written is 0 in read/write

AndreasReich committed:
Merge pull request #1018 from Unity-Technologies/unity-master-fix-tls-hang-on-connection-close-during-readwrite

[tls] Fix read/write ignoring error code and requesting more read/write calls

Commit:
Increasing sync for i2cpp debugger stack frames (case 1064723)

The thread local storage of sequence points and method execution
contexts between IL2CPP and the mono debugger code was only being
synchronized at certain times, mainly when breakpoints were processed.
This could lead to a loss of synchronization after functions are exited
and debugger frame commands accessing invalid stack data. This change
adds synchronization for these data structures right before any managed
method exit, when the method execution context for that method is
destroyed. Also optimizing memory allocations by only allocating when
the stack grows and just reusing the memory otherwise.

mderoy committed:
Merge pull request #976 from Unity-Technologies/unity-master-fix-ip-exceptions

Fix issue where loopback interface causes exception 1027045

Commit:
Merge pull request #1020 from Unity-Technologies/il2cpp-debugger-stack-frames-sync

Increasing sync for i2cpp debugger stack frames (case 1064723)

joncham committed:
Implement mono_unity_backtrace_from_context same as old mono for now (case 1074280)

joncham committed:
Get underlying method in case wrapper method is exposed to user code (case 1073634)

AndreasReich committed:
Merge pull request #1019 from Unity-Technologies/unity-master-fix-server-clientauth-error-on-no-certificate

 UnityTls server no longer throws an exception if there is no client cert during client authentication

joncham committed:
Merge pull request #1023 from Unity-Technologies/unity-master-fix-osx-backtrace

Implement mono_unity_backtrace_from_context same as old mono (case 1074280)

joncham committed:
Merge pull request #1024 from Unity-Technologies/unity-master-wrapper-methods-stack-frame

Get underlying method in case wrapper method is exposed to user code (case 1073634)

joshpeterson committed:
Don't define mmap for the clang toolchain.

The `mmap` definition problem only occurs with the GCC toolchain. With the
clang toolchain we should not redefine `mmap`, and instead use the
definition from the NDK headers.

mderoy committed:
Merge pull request #1010 from Unity-Technologies/unity-master-cleanup-thread-info

cleanup the threadinfo before shutdown

AndreasReich committed:
Fixed SslStream/UnityTlsContext not reporting gracefully closed Streams on read
Due to the way MobileAuthenticatedStream.InteralRead sets the "wouldBlock" the previous checks would falsely not report a closed stream,
which led to ReadAsync becoming stuck indefinitely
(no hangs have been observed/reported with "normal" Read, but I can't exclude that this might have happend as well with unlucky timing)

AndreasReich committed:
Merge pull request #1028 from Unity-Technologies/unity-master-fix-tls-no-graceful-close-during-read

Fixed SslStream/UnityTlsContext not reporting gracefully closed Streams on read

Commit:
Export bdwgc symbols we need to patch

Commit:
switch bdwgc repo to unity-master branch

joncham committed:
Avoid deadlock during GC. Check for GC_DONT_GC environment variable at startup rather than during collection with world stopped since g_hasenv may take OS locks and encounter deadlock. (case 1077203)

joncham committed:
Handle ref returns in runtime invoke wrappers (case 1076527)

Indirectly load ref return values in invoke wrappers and return by value.
Don't share wrappers with ref returns since indirect load is used.
Add soft debugger support for methods with ref returns.

joshpeterson committed:
Merge pull request #1027 from Unity-Technologies/android-ndk-r16

 Build Mono for Android with NDK r16b

Commit:
Don't assume an Invoke method (case 1075581)

The delegate begin invoke implementation needs to determine whether it
should invoke a method or a delegate. Previously, it assumed the
presence of a method named `Invoke` on the declaring type of the method
meant that type was a delegate. Of course, the user can name a method
`Invoke`! If that happened, the method was never really called, because
that `Invoke` method was called instead.

Commit:
Merge pull request #1030 from Unity-Technologies/export-bdwgc-symbols

Export bdwgc symbols we need to patch

joncham committed:
Revert "Get underlying method in case wrapper method is exposed to user code (case 1073634)"

This reverts commit 32ff312d2bcd3e6f6a3f01ecee46184d5b94c2a1.

joncham committed:
Merge pull request #1034 from Unity-Technologies/unity-master-revert-wrapper-methods-stack-frame

Revert "Get underlying method in case wrapper method is exposed to us…

joncham committed:
Merge pull request #1031 from Unity-Technologies/unity-master-boehm-hasenv-deadlock

Avoid deadlock during GC. (case 1077203)

joncham commented:
Where are we at on maybe merging this now? Need additional testing?

AndreasReich commented:
Probably I should prototype the Unity side of this again and see if it really has the desired effect. Questionable when I find time again for this.
I feel like there really should be isolated tests on this, but I have a hard time figuring out how to do that in Mono

AndreasReich commented:
Btw. there have been several people asking me for this feature by now, so I probably should give it a chance.
I'll check again if I find a way to write a Mono only unity test that does some native dll loading and unloading.

@bitter: Did you look at the diffs?

joshpeterson committed:
Merge pull request #1033 from Unity-Technologies/dont-assume-invoke

Don't assume an Invoke method (case 1075581)

joncham committed:
Only link bdwgc related libraries when targeting bdwgc

joncham committed:
Merge pull request #1035 from Unity-Technologies/unity-master-fix-sgen-msvc

Fix SGen build in MSVC

AndreasReich committed:
Fixed potential issue with garbage collection of delegates in UnityTls binding
Before it might have happend that the garbage collector destroys delegates that we pass to native code.

AndreasReich committed:
Merge pull request #1038 from Unity-Technologies/unity-master-fix-potential-gc-issue-with-tls-callbacks

Fixed potential issue with garbage collection of delegates in UnityTls

Commit:
Use the latest BDWGC code

joshpeterson committed:
Merge pull request #1039 from Unity-Technologies/bump-bdwgc

Use the latest BDWGC code

joncham committed:
Merge pull request #1037 from Unity-Technologies/unity-master-debugger-ref-returns

Handle ref returns in runtime invoke wrappers (case 1076527)

joncham committed:
Sync il2cpp revision 0a4be6a188cca18caa45d5baad5da037235c1f66

joncham committed:
Sync il2cpp revision 1a7f010a556c76a28ba09668920363191b8ebe29

joncham committed:
Sync il2cpp revision 37915d6211cbce909d8338c8bf5cd7a75af6343e

joncham committed:
Sync il2cpp a20072651ca630bd791734da60fc561260d1c46c

joncham committed:
Sync il2cpp revision 2e4b359d645856654955f175c11bdfc733005222

joncham committed:
Sync il2cpp revision 3176daa483ca4da5a5b38f62018b80f7b511efea

joncham committed:
Merge pull request #1040 from Unity-Technologies/unity-master-sync-il2cpp

Sync changes from IL2CPP

juj committed:
Fix html5 build by adding forward declares to referenced APIs

Commit:
Revert accidental commit

joncham committed:
Use empty array for wrapper methods (case 1073634)

Wrapper methods should not be exposed to user code but in some cases
they end up in StackFrame's. Return empty parameter array
in these cases since metadata code cannot handle wrapper methods.

mderoy committed:
pull in fix for 1066693 from upstream

mderoy committed:
Merge pull request #1042 from Unity-Technologies/unity-master-fix-ienumerable-toarray

pull in fix for 1066693 from upstream

AndreasReich committed:
Replaced size_t class in UnityTls and replaced with simple aliases to IntPtr
Since compiler handle structs differently on function calls this may have caused issues when passing arguments between managed and native code.
Even if not, we're clearly safer with the explicit casts now.

Commit:
Expose incremental gc time slice in mono API

joncham committed:
Merge pull request #1041 from Unity-Technologies/unity-master-bug-1073634

Use empty array for wrapper methods (case 1073634)

Commit:
Add mono_gc_is_incremental

Commit:
Update bdwgc

Commit:
Update bdwgc, this time with fixed version

Commit:
Fix building mono with old boehm

alexandru-unity committed:
* fixed incorrect offsetting in mono_mempool_foreach_block
* removed start/stop loop when capturing heap sections in order to prevent data inconsistencies
* moved stop/start world calls to the start/end of the capture function, we should always be stopped when collecting meta information
* added waiting for currently running GC op

Commit:
Merge pull request #1045 from Unity-Technologies/expose-incremental-gc-time-slice-

Expose incremental gc time slice

alexandru-unity committed:
removed redundant checks to GC_enable/disable, as the state is ref counted

alexandru-unity committed:
removed redundant check for gc state as this is already handled by GC_wait_for_gc_completion

Commit:
Merge pull request #1046 from Unity-Technologies/bdwgc-update

Update bdwgc

Commit:
expose GC_collect_a_little

Commit:
Merge pull request #1048 from Unity-Technologies/expose-incremental-gc-time-slice-

expose GC_collect_a_little

joshpeterson committed:
Allow the debugger to build with the Tiny runtime

These changes are mirrored from the IL2CPP repo.

realek committed:
Merge pull request #1047 from Unity-Technologies/scripting_memory_capture_crash_fixes

Scripting memory snapshot crash fixes

alexandru-unity committed:
removed call to  GC_wait_for_gc_completion(int) this call would result in a invalid declaration, furthermore incremental GC is not enabled currently on trunk

realek committed:
Merge pull request #1052 from Unity-Technologies/fix_memory_snapshot_compile_issue

Fix compilation error inside scripting_memory_capture logic

AndreasReich committed:
Merge pull request #1051 from Unity-Technologies/unity-master-fix-potential-unitytls-marshalling-issue-with-size_t

fix potential unitytls marshalling issue with size t

Commit:
Add a clang-format style file to Mono

This is a minimal clang-format file for Mono code. The code is not consistent
in its formatting, so this file is most often useful with git-clang-format [1],
which formats only changed lines.

Usage:
 * Make some changes to Mono C code or headers
 * Run `git clang-format` (maybe add -f if there are line ending issues from git)
 * Your changed code should meet the formatting guidelines for Mono

Installation:
 * Install LLVM tools (including clang-format) for your OS [2]
 * Install Pythong on your OS

Code style:
 Our goal is to match the Mono code style, not define a new one. The style
 options [3] are many. Here we try to capture only a few. If you find something
 is missing, feel free to add configuration options.

[1] https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format
[2] http://releases.llvm.org/
[3] https://clang.llvm.org/docs/ClangFormatStyleOptions.html

joshpeterson committed:
Merge pull request #1053 from Unity-Technologies/add-clang-format

Add a clang-format style file to Mono

Commit:
Use mono_gc_wbarrier_set_field instead of il2cpp_gc_wbarrier_set_field, so we can use barriers on both mono and il2cpp

joncham committed:
Update README.md for 2018.3 branches

joncham committed:
Take fast path only for delegate types not Delegate/MulticastDelegate (case 1085270)

See upstream PR: https://github.com/mono/mono/pull/11182

joncham committed:
Merge pull request #1056 from Unity-Technologies/unity-master-bug-1085270

Take fast path only for delegate types not Delegate/MulticastDelegate…

Commit:
Use mono_gc_wbarrier_generic_store instead of mono_gc_wbarrier_set_field

joshpeterson committed:
Merge pull request #1054 from Unity-Technologies/tiny-debugger-build

Allow the debugger to build with the Tiny runtime

joshpeterson committed:
Build the System.Data.DataSetExtensions assembly

The unityjit and unityaot profiles should have this assembly. It exists
in the 2.0 profile for the old scripting runtime, and we don't see any
need to remove it for the new scripting runtime.

This corrects case 1071607 in Unity.

joshpeterson committed:
Merge pull request #1060 from Unity-Technologies/add-data-set-extensions

Build the System.Data.DataSetExtensions assembly

joncham committed:
Reduce codegen overhead for debugger to single check. Avoid retrieving sequence point until actually needed.

Brings use from ~20x slower to ~5x slower when debugger codegen is enabled vs no debugger codegen.

Commit:
Use bdwgc nanosecond time limits

Commit:
Merge pull request #1061 from Unity-Technologies/high-precision-timers

Use bdwgc nanosecond time limits

mrvoorhe committed:
Additional constructor preservations

Upcoming changes to monolinker will make additional optimizations when no instance ctors are marked.

To be safe, add preservations for the ctor's of types that are created in the runtime.

The change to preserve the .ctor() of`System.Runtime.Remoting.Messaging.AsyncResult`  was made for more than out of caution.  We have a simple delegate test that hung without this.

mrvoorhe committed:
Merge pull request #1062 from Unity-Technologies/unity-master-ctor-preservations

Additional constructor preservations

joncham committed:
Use mono_mb_emit_op for CEE_LDOBJ to ensure wrapper data is added for indirect load of a ref return value. (case 1096820)

ashwinimurt committed:
Add bee script to pull down required mono-build-deps from stevedore

ashwinimurt committed:
Gitlab CI config to use Bee

ashwinimurt committed:
Checkout us and boo on build machines

ashwinimurt committed:
Use unzip installed on the machine

ashwinimurt committed:
Update Android NDK dir

ashwinimurt committed:
Update tools dir

alexandru-unity committed:
fix for case 984330, static field bytes being reported without any static fields being reported as present.

* changed class crawling to be recursive and now store non inited classes
* metadata type reporting will now also report invalid uninitialized types
* metadata type reporting will now account for array types being fields inside a class

Commit:
Move the location of mono-build-tools-extra

This repo has been moved to Github Enterprise, so point the clone code
in the build script to the new location.

alexandru-unity committed:
corrected, unecessary walking of classes already present inside the hash table.
changed code so we avoid an extra lookup

ashwinimurt committed:
Merge pull request #1064 from Unity-Technologies/use-stevedore

Add bee script to pull down required mono-build-deps from stevedore

ashwinimurt committed:
Merge pull request #1066 from Unity-Technologies/support-rearranged-mono-build-deps

Update external build-deps tools paths (MBE)

ashwinimurt committed:
Add run_all build stage

joncham committed:
Merge pull request #1063 from Unity-Technologies/unity-master-ref-return-struct

Use mono_mb_emit_op for CEE_LDOBJ to ensure wrapper data is added for indirect load of a ref return value. (case 1096820)

ashwinimurt committed:
Make build steps auto

dbuckley3d committed:
udpate bdwgc to handle suspend on altstack

dbuckley3d committed:
update bdwgc to handle suspend on altstack

alexandru-unity committed:
replaced class retrieval method as per PR comments

realek committed:
Merge pull request #1065 from Unity-Technologies/scripting_memory_capture_static_field_byte_reporting_bugfix

Scripting memory capture - static field byte reporting fix

joncham committed:
Merge pull request #1077 from Unity-Technologies/handle-suspend-on-altstack-2

update bdwgc to handle suspend on altstack

ashwinimurt committed:
Remove ono stuff

ashwinimurt committed:
Additional fixes to collect builds

joshpeterson committed:
Merge pull request #1078 from Unity-Technologies/move-mono-build-tools-extra

Move the location of mono-build-tools-extra

joncham committed:
Store method pointer, invoker indices, and reverse pinvoke indices in per assembly table rather than in metadata binary file.

This is step towards per assembly conversion.

Commit:
Make the global break check volatile

The Suspend debugger test was hanging intermittently on most platforms.
These seemed to occur because the codegen checks for
`g_Il2CppDebuggerCheckPointEnabled` were not getting the proper value,
and assuming the check point was not set, so the VM would not suspend.

This meant that the main thread could not be suspended when the debugger
was trying to exit, so the main thread kept looping forever, with out
allowing the debugger to exit it.

alexandru-unity committed:
*fixed array field names missing `[]`
*reverted filtering uninitialized classes out to avoid crashes

realek committed:
Merge pull request #1086 from Unity-Technologies/scripting_memory_capture_metadata_crawling_fix

[Memory Snapshot] Metadata crawling, use only initialized types

joncham committed:
Produce debugger information per assembly.

Collect sequence points per assembly
Retrieve sequence point information per image
Write all debugger structures to a single file for each assembly
Write codegen and debugger structures as C code. This saves ~40% compile time on MSVC

Commit:
Remove all UNITY_TINY #if statements from il2cpp-stubs.cpp

Since we run the Tiny profile on the libil2cpp backend, we don't need
any differences in this file.

Commit:
Tiny debugger tests

* Ignoring the EventSets test because it tests exceptions.
* Fixing the vm.Exit() path that doesn't use System.Environment.Exit().

ashwinimurt committed:
Add stevedore repo name to config file

ashwinimurt committed:
Merge pull request #1092 from Unity-Technologies/stevedore-repo-name

Add stevedore repo name to config file

Commit:
Remove Class From Codegen headers

ashwinimurt committed:
Merge pull request #1075 from Unity-Technologies/use-stevedore-gitlab-ci

Add Gitlab CI configs

UnityAlex committed:
Add ability to redirect mono stdout logging to unity's vprintf function.

UnityAlex committed:
Update goutput.c

Changing unity's vprintf adapters to use the va_list param instead of always passing NULL.

ashwinimurt committed:
Add option to use stevedore or ono

ashwinimurt committed:
Fix 7z path

ashwinimurt committed:
Merge pull request #1094 from Unity-Technologies/fix-7z-path

Fix 7z path

UnityAlex committed:
Adding intermediate wrap_our_vprintf wrapper to allow for the va_list to be properly setup before calling our_vprintf. Also fixed some code formatting.

UnityAlex committed:
Merge pull request #1093 from Unity-Technologies/unity-master-add-unity-vprintf-redirect

Add ability to redirect mono stdout logging to unity's vprintf function.

Commit:
Don't shutdown in mono_runtime_try_shutdown

The Mono implementation of mono_runtime_try_shutdown does not actually
shut down the VM. Instead, it sets the flags to indicate the VM is
shutting down. The VM should actually shutdown in mono_runtime_quit.

In some cases, we see this code path taken when processing the
CMD_VM_EXIT command. This seems unlikely, but can happen when the code
above cannot find a suspended managed thread. If
mono_runtime_try_shutdown actually waits for managed threads to
complete, it will hang, as the main thread won't be suspended because
resume_vm is called just above mono_runtime_try_shutdown.

So for IL2CPP mono_runtime_try_shutdown doesn't do anything, but
mono_runtime_quit will actually shutdown the VM.

Commit:
Don't shutdown in mono_runtime_try_shutdown

The Mono implementation of mono_runtime_try_shutdown does not actually
shut down the VM. Instead, it sets the flags to indicate the VM is
shutting down. The VM should actually shutdown in mono_runtime_quit.

In some cases, we see this code path taken when processing the
CMD_VM_EXIT command. This seems unlikely, but can happen when the code
above cannot find a suspended managed thread. If
mono_runtime_try_shutdown actually waits for managed threads to
complete, it will hang, as the main thread won't be suspended because
resume_vm is called just above mono_runtime_try_shutdown.

So for IL2CPP mono_runtime_try_shutdown doesn't do anything, but
mono_runtime_quit will actually shutdown the VM.

UnityAlex committed:
Adding set_vprintf_func defines back into glib.h as they were previously in old mono

UnityAlex committed:
Delete versions-win64.txt

UnityAlex committed:
Merge pull request #1096 from Unity-Technologies/unity-master-add-unity-vprintf-redirect

Fix build for missing implicit define

Commit:
Fix a build error in Tiny debugger tests

Commit:
Switch back to LARGE_CONFIG

Commit:
Switch back to LARGE_CONFIG (using master branch)

Commit:
Merge pull request #1097 from Unity-Technologies/use-large-config

Use large config

Commit:
Remove a stray # character

Commit:
Move the exit code setting back after try_shutdown

This reverts part of 78cc586d935f2009058602f2bd9adedca1442442, now that
`mono_runtime_try_shutdown` does not actually shut down, but instead
does nothing with IL2CPP.

Commit:
Shutdown the runtime properly for Tiny with the debugger

Commit:
cache pause point active state

Commit:
Inline Sequence Point checks; hard code sequence point reference into macro invocation instead of looking up from index; hard code sequence point type into macro invocation

miniwolf committed:
Fix parenthesis position in debugger-agent

joshpeterson committed:
Merge pull request #1100 from Unity-Technologies/debugger-agent/crash-on-clear-assembly

Fix parenthesis position in debugger-agent

ashwinimurt committed:
Fix typo

ashwinimurt committed:
Bokken runner for mac and android builds

ashwinimurt committed:
Remove when: manual trigger

ashwinimurt committed:
Remove manual from bareminimum

ashwinimurt committed:
Merge pull request #1095 from Unity-Technologies/use-stevedore-production

Add option to use stevedore or ono for external build deps

Commit:
Remove the <limits> and <vector> headers from being included in codegen

Commit:
Implement a ReadString overload (case 1106236)

In the `XmlSerializationReader` class, the `ReadString(string, bool)`
overload was not implemented in the Mono class library implementation used
for the unityaot profile. The implementation used for the unityjit
profile from the reference source implements this method.

I've copied the implementation from the reference source code here into
the code used for the unityaot profile.

ashwinimurt committed:
Set exe bit on bee

ashwinimurt committed:
Set using stevedoreBuildDeps as default

ashwinimurt committed:
Merge pull request #1102 from Unity-Technologies/set-exe-bit

Set exe bit on bee

Commit:
Add missing write barriers

Commit:
Better check for Boehm then using mono_gc_is_moving

joshpeterson committed:
Merge pull request #1101 from Unity-Technologies/implement-read-string

Implement a ReadString overload (case 1106236)

Commit:
Rename UNITY_TINY defines to IL2CPP_TINY

Defines in the IL2CPP code should not be named for UNITY, as we want
IL2CPP to work independently. Clean up the names of the `UNITY_TINY`
defines then by renaming them.

This change should have no impact on runtime behavior or code size.

ashwinimurt committed:
Merge pull request #1103 from Unity-Technologies/set-default-build-deps

Set using stevedoreBuildDeps as default

Commit:
Don't attempt to stop threads on exit

With IL2CPP, we don't have a good way to stop managed threads. So let's
just ignore that and exit the process.

This won't clean up and run finalizers, but the process it ending
anyway, so we don't care too much.

Commit:
Remove no longer needed debugger check code

UnityAlex committed:
Added ability to continue domain unload when an exception is encountered

UnityAlex committed:
Merge pull request #1105 from Unity-Technologies/unity-master-exception-domain-unload

Ability to continue domain unload when an exception is encountered

Commit:
Run simplified finalizer logic when gc is disabled to allow child processes to complete

joncham committed:
Fix reentrant exception handling.

Store handled status in local that is passed into
handler rather than a single thread local value
which can be overwritten in case of reentrance.

Commit:
Fix assert

Commit:
Add API to disable incremental mode

Commit:
Merge pull request #1113 from Unity-Technologies/fix-null-str-assert

Allow namespace-agnostic class lookup by passing NULL namespace.

Commit:
Update bdwgc

joncham committed:
Use gboolean rather than C++ bool type in Mono C code.

Commit:
Merge pull request #1114 from Unity-Technologies/incremental-gc-editor-support

Incremental gc editor support

ashwinimurt committed:
Report status to github

ashwinimurt committed:
Merge pull request #1116 from Unity-Technologies/report-status-to-github

Report CI status to Github

joncham committed:
Merge pull request #1115 from Unity-Technologies/unity-master-sync-il2cpp

Sync IL2CPP changes up to aebcd0f844270d5f407ec3d22f20a066d7028d36

joncham committed:
Attempt to use Win32 APIs first to retreive current timezone. Registry value can be null/corrupt (case 1076679)

joncham committed:
Merge pull request #909 from Unity-Technologies/feature/coverage

Fix up Code Coverage support to support domains

vincent-unity committed:
[Android] Fix the issue that Environment.TickCount returns wrong value due to /proc/uptime is unaccessible on Android 8 and above.

dbuckley3d committed:
Merge branch 'unity-master' of https://github.com/Unity-Technologies/mono into unity-master

dbuckley3d committed:
Tweak scripts to allow Linux to find correct binaries, config and libraries

Commit:
Don't use exceptions for time zones (case 1100856)

The current class library code uses managed exceptions for flow control
for time zone handling. Specifically, it depends on exceptions to occur
if one of the files that _might_ have time zone information does not
exist.

On Android with IL2CPP we run into problems with C++ exceptions causing
hangs in some cases. Since IL2CPP implements managed exceptions using
C++ exceptions, this code can cause problems.

This change corrects Unity case 1100856.

joshpeterson committed:
Merge pull request #1123 from Unity-Technologies/avoid-timezone-exceptions

Don't use exceptions for time zones (case 1100856)

dbuckley3d committed:
Merge pull request #1121 from Unity-Technologies/linux-mono-scripts

Linux mono scripts

joncham committed:
Merge pull request #1119 from Unity-Technologies/unity-master-reentrant-structured-exceptions

Fix reentrant exception handling (case 1093869)

joncham committed:
Merge pull request #1118 from Unity-Technologies/unity-master-fix-win32-timezone

Attempt to use Win32 APIs first to retreive current timezone (case 1076679)

UnityAlex committed:
get_shared_type now handles situation where enum types are also generic. Fixes nullref exception.

Commit:
Don't use thread pinning in Boehm

Commit:
Fix build

Commit:
Fix build

UnityAlex committed:
Merge pull request #1125 from Unity-Technologies/unity-master-fix-1111048

get_shared_type now handles situation where enum types are also gener…

Commit:
Add comment

Commit:
Merge pull request #1126 from Unity-Technologies/fix-boehm-thread-pinning

Fix boehm thread pinning

joncham committed:
Implement marshaling for generic struct fields (case 1106422)

joncham committed:
Merge pull request #1127 from Unity-Technologies/unity-master-system-numerics-sizes

Fix System.Numerics sizes (case 1118631)

joncham committed:
Merge pull request #1128 from Unity-Technologies/unity-master-ptr-to-structure-generic

Implement marshaling for generic struct fields (case 1106422)

joncham committed:
Throw ArgumentNullException if null method is passed to delegate contructor via invoke. (case 1091693)

joncham committed:
Merge pull request #1129 from Unity-Technologies/unity-master-delegate-ctor-argument-null

Throw ArgumentNullException if null method is passed to delegate contructor via invoke. (case 1091693)

lateralusX committed:
Windows support to abort blocking system calls. (#12654)

* Fix race condtion in socket altertable wait implementation on Windows.

* Add support to abort blocking Windows mono_w32file_read/write methods.

Commit:
Sync debugger agent changes from the IL2CPP repo

These changes are from

https://gitlab.internal.unity3d.com/vm/il2cpp/merge_requests/1226

joncham committed:
Grow StackSlotInfo array rather than asserting (case 1103205)

joshpeterson committed:
Merge pull request #1133 from Unity-Technologies/sync-debugger-from-il2cpp

Sync debugger agent changes from the IL2CPP repo

UnityAlex committed:
Merge pull request #1131 from Unity-Technologies/unity-master-fix-win-socket-race-condition

Windows support to abort blocking system calls. (#12654)

Commit:
Remove IL2CPP files that are now in the IL2CPP repo

These files are now in the IL2CPP repo, so remove them here to avoid
confusion.

Commit:
Sync the debugger-agent.c with the IL2CPP repo

These changes are from
https://gitlab.internal.unity3d.com/vm/il2cpp/merge_requests/1235

joncham committed:
Use wrapped method if available (case 1093274)

Reverts part of fix for case 1073634

Commit:
Don't include the IL2CPP file when building for Mono

joncham committed:
Merge pull request #1136 from Unity-Technologies/unity-master-fix-1093274

Use wrapped method if available (case 1093274)

Commit:
Add the Mono version of the common defines used in the debugger

joshpeterson committed:
Merge pull request #1135 from Unity-Technologies/il2cpp-sync-debugger-agent

 Sync the debugger-agent.c with the IL2CPP repo

joncham committed:
Merge pull request #1134 from Unity-Technologies/unity-master-fix-1103205

Grow StackSlotInfo array rather than asserting (case 1103205)

joncham committed:
Merge pull request #1120 from Unity-Technologies/unity-master-fix-android-tickcount

[Android] Fix the issue that Environment.TickCount returns wrong value…

Commit:
Fix issue where TLS requests would reallocate a buffer when it could reuse it

mderoy committed:
Merge pull request #1138 from Unity-Technologies/unity-master-tls-allocation

[Case 1084800] Fix issue where TLS requests would reallocate a buffer when it could reuse it

joncham committed:
Fix interrupting blocking file IO on Windows.

Revert part of https://github.com/Unity-Technologies/mono/pull/1131
as it causes hangs in calls to CancelSynchronousIo

Commit:
graft of upstream fix https://github.com/mono/mono/pull/9342/files for case 1070667

mderoy committed:
Merge pull request #1140 from Unity-Technologies/unity-master-dataserializer-fix

Fix serialization issue with DataContractJsonSerializer UseSimpleDictionaryFormat (case 1070667)

joncham committed:
Merge pull request #1139 from Unity-Technologies/unity-master-fix-readfile-blocking

Fix interrupting blocking file IO on Windows.

joncham committed:
Bump bdwgc submodule to get fix for GetThreadContext (case 1114668)

joncham committed:
Merge pull request #1143 from Unity-Technologies/unity-master-bdwgc-getthreadcontext

Bump bdwgc submodule to get fix for GetThreadContext (case 1114668)

Commit:
Update bdwgc

Commit:
Merge pull request #1151 from Unity-Technologies/avoid-full-gc-in-table-grow

Update bdwgc

ashwinimurt committed:
Upgrade automake to 1.16.1

ashwinimurt committed:
Merge pull request #1153 from Unity-Technologies/automake-1.16.1

Upgrade automake to 1.16.1

ashwinimurt committed:
Change stevedore repo to public

ashwinimurt committed:
Merge pull request #1154 from Unity-Technologies/change-to-public

Change stevedore repo to public

netizen539 committed:
Update bdwgc for memory snapshot crash fix

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=982) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=982) before we can accept your contribution.<br/>**4** out of **11** committers have signed the CLA.<br/><br/>:white_check_mark: joshpeterson<br/>:white_check_mark: UnityAlex<br/>:white_check_mark: AndreasReich<br/>:white_check_mark: ashwinimurt<br/>:x: Josh Peterson<br/>:x: netizen539<br/>:x: mderoy<br/>:x: lateralusX<br/>:x: vargaz<br/>:x: dbuckley3d<br/>:x: jechter<br/><hr/>**Josh Peterson** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/mono?pullRequest=982) it.</sub>
